### PR TITLE
[API] Finish covering managing orders

### PR DESCRIPTION
**Pull request by TheMilek**
| Q               | A                                                            |
|-----------------|--------------------------------------------------------------|
| Branch?         | 1.13                   |
| Bug fix?        | no                                                     |
| New feature?    | yes                                                       |
| BC breaks?      | no                                                       |
| Deprecations?   | no  |
| License         | MIT                                                          |

### CODE REVIEW

File: features/payment/managing_payments/accessing_payments_order_from_payment_index.feature [hunk 1]
```diff
@@ -12,8 +12,14 @@ Feature: Accessing payment's order from the payment index
...
-    @ui
-    Scenario: Accessing payment's order from the payments index
+    @ui @no-api
+    Scenario: Accessing payment's order from the payments index page
         Given I am browsing payments
         When I go to the details of the first payment's order
         Then I should see order page with details of order "#00000001"
+
+    @no-ui @api
+    Scenario: Accessing payment's order from the payments resource
```
Comment (Wojdylak) [inline features/payment/managing_payments/accessing_payments_order_from_payment_index.feature]: Could you connect these scenarios?
```
        Then I should see order "#00000001" with details
```

File: features/shipping/managing_shipments/accessing_shipments_order_from_shipment_index.feature [hunk 2]
```diff
@@ -15,8 +15,14 @@ Feature: Accessing shipment's order from the shipments index
...
-    @ui
-    Scenario: Accessing shipment's order from the shipments index
+    @ui @no-api
+    Scenario: Accessing shipment's order from the shipments index page
         When I browse shipments
         And I move to the details of first shipment's order
         Then I should see order page with details of order "00000001"
+
+    @no-ui @api
+    Scenario: Accessing shipment's order from the shipments resource
```
Comment (Wojdylak) [inline features/shipping/managing_shipments/accessing_shipments_order_from_shipment_index.feature]: Could you connect these scenarios?

File: src/Sylius/Behat/Context/Api/Admin/ManagingOrdersContext.php [hunk 3]
```diff
@@ -978,6 +997,77 @@ public function productSubtotalShouldBe(string $productName, string $subTotal):
...
+        Assert::true(
+            $this->responseChecker->hasValue($this->client->showByIri($shipmentIri['@id']), 'state', ShipmentInterface::STATE_SHIPPED),
+            sprintf('Shipment for this order is not %s', ShipmentInterface::STATE_SHIPPED),
+        );
+    }
+
+    /**
+     * @Then this order should have order shipping state :orderShippingState
+     */
+    public function itShouldHaveOrderShippingState(string $orderShippingState): void
```
Comment (Wojdylak) [inline src/Sylius/Behat/Context/Api/Admin/ManagingOrdersContext.php]: The scenario and function names are not consistent.

File: src/Sylius/Behat/Context/Api/Admin/ManagingPaymentsContext.php [hunk 4]
```diff
@@ -44,6 +47,34 @@ public function iAmBrowsingPayments(): void
...
+        /** @var OrderInterface $order */
+        $order = $this->iriConverter->getResourceFromIri($firstPayment['order']);
+
+        $this->client->customItemAction(Resources::ORDERS, $order->getTokenValue(), HttpRequest::METHOD_GET, 'payments');
+    }
+
+    /**
+     * @Then I should see order :order
+     */
+    public function iShouldSeeOrderPageWithDetailsOfOrder(OrderInterface $order): void
```
Comment (Wojdylak) [inline src/Sylius/Behat/Context/Api/Admin/ManagingPaymentsContext.php]:
```suggestion
    public function iShouldSeeOrder(OrderInterface $order): void
```

File: src/Sylius/Bundle/ApiBundle/Doctrine/QueryExtension/OrderExtension.php [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+        string $resourceClass,
+        string $operationName = null,
+        array $context = [],
+    ): void {
+        if (!is_a($resourceClass, OrderInterface::class, true)) {
+            return;
+        }
+
+        $user = $this->userContext->getUser();
+        if ($user instanceof AdminUserInterface && in_array('ROLE_API_ACCESS', $user->getRoles(), true)) {
```
Comment (Wojdylak) [inline src/Sylius/Bundle/ApiBundle/Doctrine/QueryExtension/OrderExtension.php]: Could we use  `SectionProviderInterface` here and check only if it is `AdminApiSection`?

File: src/Sylius/Bundle/ApiBundle/Doctrine/QueryExtension/OrderExtension.php [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+        }
+    }
+
+    /**
+     * @param array<mixed> $context
+     * @param array<mixed> $identifiers
+     */
+    public function applyToItem(QueryBuilder $queryBuilder, QueryNameGeneratorInterface $queryNameGenerator, string $resourceClass, array $identifiers, string $operationName = null, array $context = []): void
+    {
+        $this->applyToCollection($queryBuilder, $queryNameGenerator, $resourceClass, $operationName, $context);
```
Comment (jakubtobiasz) [inline src/Sylius/Bundle/ApiBundle/Doctrine/QueryExtension/OrderExtension.php]: It looks wierd as hell. It'd be better to write a private method with logic from the `applyToCollection` and invoke it in both `applyToCollection` and `applyToItem` methods.

File: src/Sylius/Bundle/ApiBundle/Doctrine/QueryExtension/OrderExtension.php [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+        QueryNameGeneratorInterface $queryNameGenerator,
+        string $resourceClass,
+        string $operationName = null,
+        array $context = [],
+    ): void {
+        if (!is_a($resourceClass, OrderInterface::class, true)) {
+            return;
+        }
+
+        if ($this->sectionProvider->getSection() instanceof AdminApiSection) {
```
Comment (jakubtobiasz) [inline src/Sylius/Bundle/ApiBundle/Doctrine/QueryExtension/OrderExtension.php]: You can use early return here. 

```
if (!$this->sectionProvider->getSection() instanceof AdminApiSection) {
    return;
}

// rest of the code
```

**Review by jakubtobiasz — CHANGES_REQUESTED** [review]

File: features/order/managing_orders/order_details/being_unable_to_see_cart.feature [hunk 8]
```diff
@@ -10,7 +10,7 @@ Feature: Being unable to see details of a cart
         And the customer added "PHP T-Shirt" product to the cart
         And I am logged in as an administrator
 
-    @ui
+    @ui @api
     Scenario: Seeing basic information about an order
```
Comment (NoResponseMate) [inline features/order/managing_orders/order_details/being_unable_to_see_cart.feature]:
```suggestion
    Scenario: Being unable to see the details of a cart
```

File: features/payment/managing_payments/accessing_payments_order_from_payment_index.feature [hunk 9]
```diff
@@ -12,8 +12,8 @@ Feature: Accessing payment's order from the payment index
...
         And I am logged in as an administrator
 
-    @ui
-    Scenario: Accessing payment's order from the payments index
+    @ui @api
+    Scenario: Accessing payment's order from the payments resource
         Given I am browsing payments
         When I go to the details of the first payment's order
-        Then I should see order page with details of order "#00000001"
+        Then I should see order "#00000001" with details
```
Comment (NoResponseMate) [inline features/payment/managing_payments/accessing_payments_order_from_payment_index.feature]:
```suggestion
        Then I should see the details of order "#00000001"
```

File: features/payment/managing_payments/accessing_payments_order_from_payment_index.feature [hunk 10]
```diff
@@ -12,8 +12,8 @@ Feature: Accessing payment's order from the payment index
         And there is an "#00000001" order with "Apple" product
         And I am logged in as an administrator
 
-    @ui
-    Scenario: Accessing payment's order from the payments index
+    @ui @api
+    Scenario: Accessing payment's order from the payments resource
```
Comment (NoResponseMate) [inline features/payment/managing_payments/accessing_payments_order_from_payment_index.feature]:
```suggestion
    Scenario: Accessing payment's order from the payment
```
or just
```suggestion
    Scenario: Accessing the order of a payment
```
similar w/ shipments

File: src/Sylius/Behat/Context/Api/Admin/ManagingOrdersContext.php [hunk 11]
```diff
@@ -93,6 +96,22 @@ public function iSpecifyFilterDateFromAs(string $dateTime): void
         $this->client->addFilter('checkoutCompletedAt[after]', $dateTime);
     }
 
+    /**
+     * @When specify its tracking code as :trackingCode
+     */
+    public function specifyItsTrackingCodeAs(string $trackingCode): void
+    {
+        $this->sharedStorage->get('order')->getShipments()->first()->setTracking($trackingCode);
```
Comment (NoResponseMate) [inline src/Sylius/Behat/Context/Api/Admin/ManagingOrdersContext.php]: No. Use the API.

File: tests/Api/Admin/OrdersTest.php [hunk 12]
```diff
@@ -65,8 +65,8 @@ public function it_gets_orders_for_customer(): void
         $fixtures = $this->loadFixturesFromFiles([
             'authentication/api_administrator.yaml',
             'channel.yaml',
-            'customer.yaml',
-            'customer_order.yaml',
+            'order/customer.yaml',
+            'order/fulfilled_order.yaml',
```
Comment (NoResponseMate) [inline tests/Api/Admin/OrdersTest.php]:
```suggestion
            'order/fulfilled.yaml',
```

**Review by NoResponseMate — CHANGES_REQUESTED** [review]

File: src/Sylius/Behat/Context/Api/Admin/ManagingOrdersContext.php (hunk 11 above)
Comment (TheMilek) [reply src/Sylius/Behat/Context/Api/Admin/ManagingOrdersContext.php]: forgive me, totally forgot to change it XD

File: src/Sylius/Bundle/ApiBundle/Doctrine/QueryExtension/OrderExtension.php [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+
+        if (!$this->sectionProvider->getSection() instanceof AdminApiSection) {
+            return;
+        }
+
+        $stateParameter = $queryNameGenerator->generateParameterName('state');
+        $rootAlias = $queryBuilder->getRootAliases()[0];
+
+        $queryBuilder
+            ->andWhere(sprintf('%s.state != :%s', $rootAlias, $stateParameter))
```
Comment (NoResponseMate) [inline src/Sylius/Bundle/ApiBundle/Doctrine/QueryExtension/OrderExtension.php]: Won't this fail if we pass more than one state?

File: src/Sylius/Behat/Context/Api/Admin/ManagingPaymentsContext.php [hunk 14]
```diff
@@ -44,6 +47,34 @@ public function iAmBrowsingPayments(): void
...
+    public function iShouldSeeOrderWithDetails(OrderInterface $order): void
+    {
+        Assert::true(
+            $this->responseChecker->hasItemWithValue(
+            $this->client->getLastResponse(),
+            'order',
+            $this->sectionAwareIriConverter->getIriFromResourceInSection($order, 'admin'),
+        ),
+            sprintf('Order with number %s does not exist', $order->getNumber()),
+        );
```
Comment (NoResponseMate) [inline src/Sylius/Behat/Context/Api/Admin/ManagingPaymentsContext.php]: The formatting here is weird, but let's leave it for the ecs build to take care of.

File: src/Sylius/Bundle/ApiBundle/Resources/config/services/extensions.xml [hunk 15]
```diff
@@ -129,6 +129,15 @@
             <tag name="api_platform.doctrine.orm.query_extension.item" />
         </service>
 
+        <service id="Sylius\Bundle\ApiBundle\Doctrine\QueryExtension\OrderExtension">
+            <argument type="service" id="Sylius\Bundle\CoreBundle\SectionResolver\SectionProviderInterface" />
+            <argument type="collection" >
+                <argument type="constant">Sylius\Component\Order\Model\OrderInterface::STATE_CART</argument>
+            </argument>
```
Comment (NoResponseMate) [inline src/Sylius/Bundle/ApiBundle/Resources/config/services/extensions.xml]: Might be worth to parametrize this

**Review by NoResponseMate — APPROVED** [review]

**Review by jakubtobiasz — APPROVED** [review]

**Comment by jakubtobiasz** [comment]
Thank you, @TheMilek!

**Pull request merged**